### DISCUSSION TITLE
Using ref.read to initialize provider

### DISCUSSION THREAD

**Discussion by Velkamhell97** [rrousselGit/riverpod#2971] [Q&A]
Riverpod docs says it is a bad practice to use `ref.read` in the `build()` method, but i want to read values from one provider to initialize other provider and then listen the changes of the first provider to update the second provider, let me show for better understanding

```dart
class FollowedTopicsNotifier extends AsyncNotifier<List<FollowableTopic>> {

  @override
  FutureOr<List<FollowableTopic>> build() async {
    // initial values to initialize this provider
    final user = ref.read(userNotifier);
  
    /// I suscribe to the changes with ref.listen instead ref.watch to prevent rebuild in this provider
    ref.listen(userNotifier.select((user) => user.followedTopics), (_, topicsIds) { 
      final updatedList = state.value!.map((value) => value.copyWith(isFollowed: topicsIds.contains(value.topic.id)));
      state = AsyncData(updatedList.toList());
    });
  
    // With ref.wath the build method is called again and then the request is made again
    final topics = await ref.watch(topicsRepository).fetchTopics();
   
    final followedTopics = topics.map(
      (topic) => FollowableTopic(
        topic: topic,
        isFollowed: user.followedTopics.contains(topic.id)
      )
    ).toList();

    return followedTopics;
  }

}
```

I can't use `ref.watch` because this causes a rebuild in the current provider and therefore the state will be reset and the network request will be made again. I want to read the initial value of `userNotifier` to initialize this notifier and then suscribe to changes in `userNotifier` without causes a rebuild preventing the request from trigger again. If the `build()` method doesn't have  any `ref.watch` it is only be triggered one time, so i don't know if it is still a bad practice. If it is i would appreciate any advice and how i could achieve the desired behaviour

Thank you very much!

**Comment by rrousselGit** [comment]
Use then.listen then if you don't want updates in a notifier. Or `ref.watch(provider.select)`

Not ref.read

**Reply by Velkamhell97** [reply]
The problem with `ref.listen` is that it doesn't return the state value, and i need this value to initialize the provider, in the previous example to find the selected user topics. Even the `ref.listen` could be ommited for this answer. The main problem is when i try to read the userNotifier instance, if i use `ref.watch`, when the user adds a new selected topic, the `userNotifier ` will change and the build method of `FollowedTopicsNotifier` wil be called again, triggering the network request, even if i use the `ref.select` when the user selected topics changes the `build` method will be called too. I want to trigger the `build` method only once, to load the initial data (that dependes of `userNotifier` current value) for that reason y can't user `ref.watch` or `ref.watch(provider.select)`.

I thought about use a ProviderScope to pass this initial value in the `FollowedTopicsNotifier` constructor but i don't know if it is a good approach.

**Reply by rrousselGit** [reply]
Listen does return the current value. The subscription returned can be used to read providers
